rse-py-package.Rmd

```diff
@@ -0,0 +1,1460 @@
+What do you get?
+2. Add the statement `print(__name__)` to the end of the file `zipf.py`
+and from the python interpreter, run `import zipf`.  What is the name
+printed?
+
+### Exercise: Turn `use.py` into a unit test
+
+1. Turn the `use.py` script into a `pytest` unit test; name the file `test_zipf.py`.
+
+## How do I create a package? {#rse-py-package-creating}
+
+Let's say we wanted to add some capabilities to what
+we've built: we want to be able to generate noisy
+Zipf distributions -- noise within the same default `RELATIVE_ERROR` --
+and we want to add a Zipf-checker that can handle
+unsorted inputs:
+
+<!-- src="rse-py-package/03a/zipf.py" -->
+```python
+...as before...
+
+def make_zipf(length):
+    assert length > 0, 'Zipf distribution must have at least one element'
+    result = [1/(1 + i) for i in range(length)]
+    return result
+
+def make_noisy_zipf(length, rel=RELATIVE_ERROR):
+    data = make_zipf(length)
+    minnoise = 1.0 - rel/2
+    maxnoise = 1.0 + rel/2
+    for i in range(length):
+        data[i] = data[i]*random_uniform(minnoise, maxnoise)
+    return data
+
+def is_zipf(hist, rel=RELATIVE_ERROR):
+    assert len(hist) > 0, 'Cannot test Zipfiness without data'
+    scaled = [h/hist[0] for h in hist]
+    perfect = make_zipf(len(hist))
+    return scaled == approx(perfect, rel=rel)
+
+def is_unsorted_zipf(hist, rel=RELATIVE_ERROR):
+    sortedhist = sorted(hist)
+    return is_zipf(unsorted, rel)
+
+...as before...
+```
+
+As our package grows,
+we should split its source code into multiple files,
+and this is now starting to look like two broad types of 
+related functionality: generating a Zipf distribution, and
+checking against a Zipf distribution.  But they are related;
+and what's more, there are constants that would be common
+across the two pieces - the `RELATIVE_ERROR` constant.
+
+If we want multiple files, but they are related and
+and have some common items, that's a good indication that
+we are outgrowing a single-file module and we want a python
+package.
+
+A package in python is no more than a directory containing
+python code, and by convention a python file with a special
+name, `__init__.py` (this was a requirement before Python 3.3).
+Just as importing a module file executes the code in the module,
+importing a package executes the code in `__init__.py`.
+
+### Exercise: Create a package for our Zipf code.
+
+We're going to create a package for the code we've written
+for the Zipf distribution.  In the long and honourable(?)
+tradition of package names ending in "py", we'll call our
+package zipfpy.
+
+1. Create a `zipfpy` directory and move the file `zipf.py` into that directory.
+2. Create an empty file `zipfpy/__init__.py`.
+3. Modify the test suite so that it runs; functions will now be found in the namespace `zipfpy.zipf` rather than `zipf`.
+4. Run the test suite with `pytest`
+
+FIXME: should introduce the name 'namespace' earlier if it doesn't come up in earlier sections
```

Comment:
I think this is the first use of "namespace" - add it to the glossary as well please.

Reply:
Done